### DISCUSSION TITLE
Teslamate go forward

### DISCUSSION THREAD

**Discussion by Dulanic** [teslamate-org/teslamate#3416] [General]
Hello everyone! As we know @adriankumpf  is ceasing development of this fantastic app going forward.  With that said, we need to determine who if anyone can take more responsibility for this project going forward. While I myself am great on the SQL side, I know next to nothing about elixir.

I'd like to see who if anyone can step up to assist with which portions of this project.  If someone wants to totally convert this over to python, I can help out more! Kidding.... I think.

Some ideas:

Dashboard/SQL - @Dulanic @DrMichael , others?
Elixir - ??
Other portions I haven't thought of: ?

If you would like to assist, please speak up here and we can start to focus on how this project will continue going forward. If anyone else would like to take this discussion over, let me know! :)

I am going on vacation in a few days and won't have a lot to say until December.

**Comment by parkr** [comment]
I'd be happy to take a look at Elixir PRs and help with general administration if you still need volunteers.

**Comment by cwanja** [comment]
I will _generally_ be around. Elixir took me about five tries to spell correctly. I can review code today and add / modify for new features. But it will be buggy as hell. Regardless, will be active in discussions, issues and around. Better in Grafana / SQL.

**Comment by parkr** [comment]
I'd nominate @swiffer for Elixir review if he's interested, based on the excellent PR https://github.com/teslamate-org/teslamate/pull/3398.

**Reply by swiffer** [reply]
Thanks @parkr - happy to contribute with reviews, testing, issue grooming, releases, tooling, sql & grafana but sadly cannot do much when it comes to elixir. I've 0 experience.

**Comment by brianmay** [comment]
My observation is the MQTT support in Elixir is in a very sad state. [Tortoise](https://github.com/gausby/tortoise) which Teslamate uses, has not received any updates in the last two years, and is stuck on MQTT 3.1 with MQTT 5 support being very experimental.

There is a [fork](https://github.com/smartrent/tortoise311/) but I think it is bug fixes only. But maybe we should  seriously considering switching to that that anyway.

The Erlang [emqtt](https://github.com/emqx/emqtt/) library is better, but needs extra code to ensure the connection stays up (I have written and [published](https://github.com/brianmay/mqtt_potion/) such code), and has various support issues also (such as the hex.pm version is old, problems sending keep alives frequently enough not to get disconnected by the server, etc).

And I have seen issues in TeslaMate where the Elixir process can die (and auto restart) because something unexpected/untested happened, e.g. the Tesla API returns an unexpected response. Which might be OK, but doesn't fill me with confidence. When the process dies, that means all data that the process was keeping track of is lost.

At the risk of becoming unpopular, wondering if anybody else would be interested in participating with a complete rewrite of this project in another language. Rust is my preferred language for the task. It has strong compile time checking, which will lead to higher confidence the code is working as designed. I think this is very important for Teslamate. Ideally this would support the same database schema, same MQTT protocol, etc.

There are some things really like about Elixir, e.g. live views, etco. But I think these aren't really key features for teslamate.

(and I notice that @adriankumpf's tesla_auth app is written in Rust also...)

I don't have a huge amount of time, but do have minimal working rust code to talk to Tesla API: https://github.com/brianmay/robotica-rust/blob/main/robotica-backend/src/services/tesla/api.rs (most important missing part is streaming support - seeing as I currently use Teslamate for that...).

**Reply by ngardiner** [reply]
Why not start a brand new project at that point? The backend doesn't matter to users really, you could just keep the same db schema and grafana will render it the same. 
Rewriting TeslaMate just means you are retiring the elixir-based platform, not providing an alternative.

**Reply by swiffer** [reply]
Elixir seems to be the part where based on feedback of current volunteers there is the biggest gap.

Based on the changes Tesla is currently doing I would sit back a month or two and see how 3rd party apps, fleet-telemetry etc. are going to develop and if there is someone with enough knowledge in elixir to migrate over (if needed).

If we are required to do bigger changes to the codebase it's hard to drive these currently.

**Reply by bassmaster187** [reply]
I was asked to write a few words about the current situation. 
I am the founder of Teslalogger and as many of you know Teslamate was an 1:1 copy of Teslalogger at the beginning. Therefore, many concepts / SQL schema are just a copy of Teslalogger. And we have an import Tool to migrate your data to Teslalogger:
https://github.com/bassmaster187/TeslaLogger/tree/master/Teslamate-Import

We are 3 main developers now and the project is written mainly in c#. A very popular language and even you have no clue about it, it is possible to learn it within a couple of days if you are familiar with java. So even if I will lose the passion for this project, at least the two other developers will continue it or anybody else. I've got two Teslas and maybe will purchase a new Tesla by the end of the year, so I'm far away from leaving the project. 

https://github.com/bassmaster187/TeslaLogger

So, before you start the whole Teslamate project from scratch, consider there is a ready to go project with much more features, more developers, popular Language, 6000 users and much more. But consider as well Adrian could be ill, in vacation or has any other problems which are now more important than an hobby project.

**Reply by brianmay** [reply]
Does TeslaLogger support MQTT? That is really an essential requirement for me. And my experience with MQTT is that it is often poorly supported on unusual languages. And I think C# is hardly a popular language for IOT.

For me it might be a case of switching from one relatively obscure language to another relatively obscure language. Which doesn't help me much.

Yes, I know C# really is a popular language, but not so much for Linux or IOT development. I am not even sure right now if it is possible to develop/run C# code on nixos, my Linux based operating system I use. I am guessing it should be possible, just not familiar.

That is just my immediate thoughts right now, subject to change of course :-)

**Reply by brianmay** [reply]
I think I said this elsewhere: I am very grateful for all the work @adriankumpf has put into this project. I think everyone is. He is free to step down for any reason (as he appears to be doing right now), he is not under any obligation to explain or justify his reasons to us.

**Reply by DrMichael** [reply]
I switched from teslalogger to teslamate some years ago... :-)
And, yes, we should not forget, that @bassmaster187 came up with the great idea and initial implementation.
When comparing teslalogger and teslamate I would see the following points (I did not follow the teslalogger implementation in the last time though)

- Beauty: I agree with Christian, that teslalogger has some nicer and very beautiful dashboards. Comes partially at the cost of adapted code of panels, which makes it necessary to take on the changes of the upstream code.
- Data storage: having all data in the database appears to be cleaner and easier to migrate to me. So, switching the underlying hardware is just a matter of backup and retore of the database.
- Programming language: Right, C# is a very popular language, but as @brianmay pointed out, probably not under Linux. On the other hand Elixir is quite exotic and definitely does not have a wider programmer's base. (I personally would have preferred Java or Python.)

So, both are great pieces of software. My personal preference is teslamate just because of the data storage aspect.

**Reply by JakobLichterfeld** [reply]
There are alternatives to TelsaMate, but it is still the best choice from my point of view. Why? Regarding TeslaLogger, my opinion did not changed: https://tff-forum.de/t/teslamate-erfahrungs-und-datenthread/84179/45

**Reply by bassmaster187** [reply]
@brianmay : Yes MQTT is fully supported in Teslalogger. C# is also fully supported on Windows, Linux an MacOS. One Developer is using MacOS, but I never tried Visual Studio on Linux. All users are running the project on Debian. 

https://visualstudio.microsoft.com/de/downloads/

@DrMichael just the Geofences are not in the database, because it is easier for users to copy and paste fast chargers and make a pull requst for the fast chargers statistics. https://teslalogger.de/charging_locations.php 

https://github.com/bassmaster187/TeslaLogger/blob/master/TeslaLogger/bin/geofence.csv

From my point of view the main advantage of Teslamate over Teslalogger is the much better documentation. The main developers are loving coding and not making fancy manuals. :-)  We would love if anybody wants to write a better manual.

**Reply by BillGOH** [reply]
[Teslamateapi  ](https://github.com/tobiasehlert/teslamateapi)and the popular [TeslaSolarCharger ](https://github.com/pkuehnel/TeslaSolarCharger) depend on TeslaMate, which is more "elegant" and "slim" in comparison to Teslalogger.

**Reply by JakobLichterfeld** [reply]
The main advantage of TeslaMate is not only the error security and self-healing and SW architecture and modern SW patterns and much more, but not sharing any data with the outside, as TeslaLogger always did and still does.

**Reply by brianmay** [reply]
Key question: Is it possible to run both Tesla Mate and Tesla Logger at the same time? If so, this would considerably open up my options. I could test out Tesla Logger and not mess up my Tesla Mate setup. Or I could run Tesla Mate, while developing my own solution.

I have a vague recollection that the Tesla API only supports streaming to one client at a time though. If this is true, and if this is still the case, the answer would be a no.

Wonder how hard it would be to write some sort of Tesla API simulator for testing. This might be good for any testing of any of the loggers.

But my key concern is Tesla. e.g. on some forums I have read there is speculation that Tesla could put their API behind a paywall. Which somewhat limits the time I want to spend on this.

**Reply by ngardiner** [reply]
You can, I've run them side by side for a year now

**Reply by Adminius** [reply]
Yeah... and this is the the problem, it works only with TeslaMate and also "one man show"... 
Thats why I using EVCC now, more popular with more supported hadware. TeslaLogger will become EVCC integration this days (e.g. Session Price, PV %)

**Reply by JakobLichterfeld** [reply]
I am happy for you. Have fun posting in other projects.

**Comment by jlestel** [comment]
Hi, if you still need maintainers, I'm interested in helping you move forward, I do a lot of DevOps.

I must surely be able to help with:
- the review of the Elixir PRs
- CI and build optimization (at present I find it annoying not to have release candidates to be able to test particular versions on a large scale)
- I manage several hundred Teslamate installations so I can also help deploy versions and quickly get feedback on behavior

Indeed, maintaining the Elixir code seems to be the most complex part and I am not an expert.

Finally, I can offer financial support if there is a need for licenses or others.

PS: I also have access to the Fleet API etc. I am thinking about integrating Teslamate into this new ecosystem.

**Comment by DrMichael** [comment]
So we have:

- Dashboard/SQL - @Dulanic @DrMichael @cwanja @swiffer @jheredianet 
- Elixir Programming - @NirKli 
- Elixir Review - @jlestel @swiffer @parkr @swiffer @jsight
- Docker - @jheredianet @nelsongraca 
- CI and build: @jlestel 
- Home automation - @nelsongraca 
- General - @JakobLichterfeld

**Reply by swiffer** [reply]
@DrMichael - Out for Elixir programming - in for Everything else, sorry!

**Reply by DrMichael** [reply]
List updated...

**Reply by Dulanic** [reply]
Thanks for putting this together @DrMichael. As I thought so far, elixir is our biggest steppingstone. But we can give it some time and see what else we have.

**Reply by JakobLichterfeld** [reply]
I'm up for anything that helps. DevOps, documentation, code reviews. 
I speak many programming languages, but currently no real Elixir, but that may change.

**Reply by NirKli** [reply]
> So we have:
> 
> 
> 
> - Dashboard/SQL - @Dulanic @DrMichael @cwanja @swiffer 
> 
> - Elixir Programming - 
> 
> - Elixir Review - @jlestel @swiffer @parkr @swiffer 
> 
> - CI and build: @jlestel 
> 
> 
> 
> What about @JakobLichterfeld @jsight @NirKli ?

I would like to try to fit in elixir programming

**Reply by DrMichael** [reply]
Updated. Thanks

**Reply by DrMichael** [reply]
@jheredianet where should I put you in?

**Reply by jheredianet** [reply]
Grafana Dashboards/SQL and Docker containers

**Reply by DrMichael** [reply]
Appreciated, updated...

**Comment by nelsongraca** [comment]
volunteering for Docker, CI/CD, home assistant, and general project management stuff, can try to review PRs but I know nothing about elixir as java is my expertise

**Reply by DrMichael** [reply]
Guys help me: what is CI/CD?
I would like to keep building on one hand with @jlestel ...

**Reply by jlestel** [reply]
CI/CD is a method to frequently deliver apps to customers by introducing automation into the stages of app development. The main concepts attributed to CI/CD are continuous integration, continuous delivery, and continuous deployment.

The CI is what you see here: https://github.com/teslamate-org/teslamate/actions

As you can see, last commit on master is not passing Elixir CI

**Comment by kimmilde** [comment]
So there have been a lot of merging now. is the docker image from the documentation still valid? Is that to keep in my docker-compose?

**Reply by jlestel** [reply]
Docker-compose is on latest. This is still valid.

**Comment by DrMichael** [comment]
I'd like to test PR's in advance. How would I get an image locally, where I could make changes and run with that?

**Reply by jlestel** [reply]
Indeed, before merging a PR into a release candidate branch or into master, it would be interesting to be able to directly test the image of a branch.

I was thinking more of the following operation: when a branch is mergable and passes the auto tests => we build the image of this branch and push it to docker hub.

Example result image: teslamate/teslamate:my_feature_to_test

You could thus directly test it on your local install by changing the version in your docker-compose.

Open to discuss the git flow if you have ideas.

**Reply by jheredianet** [reply]
I second @jlestel idea. In this way it could be tested by more people at the same time without waiting just a few ones to report possible issues.

**Reply by swiffer** [reply]
not sure if it's great to have all these branches publicly available on docker hub - we will end up with a lot of different tags. do we need to test each PR individually or could we go with a :edge tag instead (that exists currently, right?)

**Reply by jlestel** [reply]
We don't necessarily need to test the image of all the PRs but I suppose we will have several to test in parallel. 
And if you want to test several branches at the same time, the edge tag is limiting atm.

When a branch is finally merged, we can also delete its tag (in the same way that we should delete the branch when it is merged).

To avoid disrupting the current docker hub, it is also possible to push these branch images into another hub (eg: teslamate/teslamate-dev).

**Reply by JakobLichterfeld** [reply]
I like the idea. I have done it in the past with artifacts, see [here for example](https://github.com/TheRealKasumi/NikoLight/blob/main/.github/workflows/nikolight-update-packaging-tool-and-nup-file-build.yml). We can extend it to build a docker image that you can download for testing.

Advantage: no one has to build docker images on their own, more can test. And reproducible. Auto delete after retention days.

**Reply by nelsongraca** [reply]
I would just create one for each PR, they can be deleted on PR merge.

**Reply by JakobLichterfeld** [reply]
Yeah, that's exactly the Plan, as you can see in my example (push on main or for all releases, and all PR to main

**Reply by micves** [reply]
That would be great. I have #3262 and several different people/platforms/architectures eager to test :)

**Reply by jlestel** [reply]
Here is the implementation I had in mind: https://github.com/teslamate-org/teslamate/pull/3432

There must be a right to be set at the org level: buildx failed with: ERROR: denied: installation not allowed to Write organization package

I don't have access to it yet.

@micves we should test it soon: ghcr.io/teslamate-org/teslamate:pr-3262 ;)

**Reply by JakobLichterfeld** [reply]
Further discussions in #3432.

**Comment by DrMichael** [comment]
> Here is the implementation I had in mind: #3432

Sounds good!

> There must be a right to be set at the org level: buildx failed with: ERROR: denied: installation not allowed to Write organization package
> I don't have access to it yet.

Hmmm, what needs to be done on org level? I cannot add members there, probably only the owner can. That is @adriankumpf ...

**Reply by adriankumpf** [reply]
@DrMichael @jlestel I've granted actions within teslamate repository write access to the teslamate package. Please try again.

**Comment by JakobLichterfeld** [comment]
All PR are labeled now, quick wins were merged.

**Reply by DrMichael** [reply]
Thank you. Excellent work!
Perhaps we clean up the discussions also...

**Comment by cwanja** [comment]
Might be opening a can of worms here, but:
1. Do we have a central chat place for us? Discord, WhatsApp, Signal, etc.? 
2. Do we have a central place for us document enhancements, changes, etc. Know @brianmay mentioned in other threads he has active items on his list. 
3. Do we have a target release? There has been lots of pushes recently to master and figure incremental releases are better than big bangs. 

This is the most activity I have seen in TeslaMate since 2019 / 2020 and am super excited. But want to make sure it is methodically done.

**Reply by NirKli** [reply]
> Might be opening a can of worms here, but:
> 
> 1. Do we have a central chat place for us? Discord, WhatsApp, Signal, etc.? 
> 
> 2. Do we have a central place for us document enhancements, changes, etc. Know @brianmay mentioned in other threads he has active items on his list. 
> 
> 3. Do we have a target release? There has been lots of pushes recently to master and figure incremental releases are better than big bangs. 
> 
> 
> 
> This is the most activity I have seen in TeslaMate since 2019 / 2020 and am super excited. But want to make sure it is methodically done. 

+1 for the central chat

**Reply by LelandSindt** [reply]
Discord ...
https://github.com/teslamate-org/teslamate/discussions/198

**Reply by LelandSindt** [reply]
... That said... +1 for keeping things open and organized here in GitHub.

**Reply by JakobLichterfeld** [reply]
Yes, we have a central place: Since we are an open source project, we track everything here on GitHub. Either in issues or here in discussions. Personally, I prefer issues for feature ideas, bugs, etc., but that's a matter of taste.

I have already drafted the next release and voted to jump directly to 1.28.0 to show that it is the first release of teslamate-org. We just finished building GHCR containers for each PR to test things. We just need a setup from Adrian, see #3481.
I will add the latest PR in the Changelog and release notes as well and release as soon as it is tested.

**Reply by brianmay** [reply]
Github also supports wiki pages (looks like currently disabled for this project). Discussions is good for meeting and making decisions, wiki would be good for posting the results (with links as appropriate) of such discussions so that we don't forget what we decided. I think searching for old decisions in chat may be difficult after a while.

**Reply by markusdd** [reply]
Yes. I am also in favor of leveraging Github Discussions and maybe the wiki.
Also Github offers Github pages where one could even build a small webpage explaining the roadmap etc.

I strongly vote against Discord. This app is a closed-source resource hog in the hands of a chinese company. No thx.

**Reply by brianmay** [reply]
If we really wanted a instant chat system, I would vote for [Matrix](https://matrix.org/) instead of Discord. A number of open source projects I am involved with already use it. I personally use the [element client](https://element.io/), but there are others.

But if we are happy with what Github provides, even better.

**Reply by nelsongraca** [reply]
I only see benefit in a chat system for end user support, for that Discord would be my choice as it seems to be the easiest one.
For dev talk we could leverage discussions here

**Comment by jheredianet** [comment]
@DrMichael I've opened this topic https://github.com/teslamate-org/teslamate/discussions/3513

Let me know what you think. It's something we should consider when approving the PRs.

**Reply by DrMichael** [reply]
Not sure what the question is. I am sure @swiffer will take care of this in https://github.com/teslamate-org/teslamate/issues/3438
And yes, I think, we should not have separate dashboards for LFP and other batteries. So, unifying in those three cases would be good.

**Comment by JakobLichterfeld** [comment]
Follow-up, two years on: the direction discussed here has been decided and is documented in #5558 — modular data sources with an explicit event contract, and the core moving to Rust. That thread covers what is decided, what is still open, and how it will be sequenced. If you're subscribed here because you care about where the core is going, that's the thread to follow now.
